src/modules/control_allocator/ControlAllocator.cpp

```diff
@@ -0,0 +1,501 @@
+	actuator_min(9) = _param_ca_act9_min.get();
+	actuator_min(10) = _param_ca_act10_min.get();
+	actuator_min(11) = _param_ca_act11_min.get();
+	actuator_min(12) = _param_ca_act12_min.get();
+	actuator_min(13) = _param_ca_act13_min.get();
+	actuator_min(14) = _param_ca_act14_min.get();
+	actuator_min(15) = _param_ca_act15_min.get();
+	_control_allocation->setActuatorMin(actuator_min);
+
+	// Maximum actuator values
+	matrix::Vector<float, NUM_ACTUATORS> actuator_max;
+	actuator_max(0) = _param_ca_act0_max.get();
+	actuator_max(1) = _param_ca_act1_max.get();
+	actuator_max(2) = _param_ca_act2_max.get();
+	actuator_max(3) = _param_ca_act3_max.get();
+	actuator_max(4) = _param_ca_act4_max.get();
+	actuator_max(5) = _param_ca_act5_max.get();
+	actuator_max(6) = _param_ca_act6_max.get();
+	actuator_max(7) = _param_ca_act7_max.get();
+	actuator_max(8) = _param_ca_act8_max.get();
+	actuator_max(9) = _param_ca_act9_max.get();
+	actuator_max(10) = _param_ca_act10_max.get();
+	actuator_max(11) = _param_ca_act11_max.get();
+	actuator_max(12) = _param_ca_act12_max.get();
+	actuator_max(13) = _param_ca_act13_max.get();
+	actuator_max(14) = _param_ca_act14_max.get();
+	actuator_max(15) = _param_ca_act15_max.get();
+	_control_allocation->setActuatorMax(actuator_max);
+
+	// Control effectiveness
+	matrix::Matrix<float, NUM_AXES, NUM_ACTUATORS> B;
+
+	switch (_param_ca_airframe.get()) {
+	default:
+		// Mock-up effectiveness matrix
+		const float B_quad_w[NUM_AXES][NUM_ACTUATORS] = {
+			{-0.5717536f,  0.43756646f,  0.5717536f, -0.43756646f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f},
+			{ 0.35355328f, -0.35355328f,  0.35355328f, -0.35355328f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f},
+			{ 0.28323701f,  0.28323701f, -0.28323701f, -0.28323701f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f},
+			{ 0.f,  0.f,  0.f,  0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f},
+			{ 0.f,  0.f,  0.f,  0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f},
+			{-0.25f, -0.25f, -0.25f, -0.25f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f, 0.f}
+		};
+		B = matrix::Matrix<float, NUM_AXES, NUM_ACTUATORS>(B_quad_w);
+		break;
+	}
+
+	// Set 0 effectieness for actuators that are disabled (act_min >= act_max)
+	for (size_t j = 0; j < NUM_ACTUATORS; j++) {
+		if (actuator_min(j) >= actuator_max(j)) {
+			for (size_t i = 0; i < NUM_AXES; i++) {
+				B(i, j) = 0.0f;
+			}
+
+		}
+	}
+
+	// Assign control effectiveness matrix
+	_control_allocation->setEffectivenessMatrix(B);
+}
+
+void
+ControlAllocator::update_allocation_method()
+{
+	int method = _param_ca_method.get();
+
+	if (_allocation_method_id != method) {
+		// Save current state
+		ControlAllocation *tmp = nullptr;
+		matrix::Vector<float, NUM_ACTUATORS> actuator_sp = _control_allocation->getAllocatedActuator();
+
+		// try to instanciate new allocation method
+		switch (method) {
+		case 0:
+			tmp = new ControlAllocationSimple();
```

Comment:
@jlecoeur I assume eventually this could be a list of allocators correct? E.g. for a VTOL we would need the multirotor allocator plus allocators for control surfaces and forward thrust.

Reply:
Eventually I would like to have a single matrix and a single allocator for all actuators. There is no real reason to treat them separately.

A blocker is the current multirotor mixer which is very specific to multirotors. A solution is to add a new child class `ControlAllocationHybrid` that internally uses two allocators for rotors and flaps.

However, if we manage to express the current multirotor mixer more generically, like with a priority vector instead of hard-coded thrust boosting, that intermediate step is not needed.

@bresch do you think we can transition the multitoror mixer to something more generic, namely not making the assumption that all actuators are effective on -Z thrust?

Reply:
@jlecoeur Currently, I don't see a reason why we could not. The desaturation logic in the multicopter mixer uses one torque / thrust axis at a time in order to un-saturate actuators.
As long as we know the priorities I don't think we are restricted to what we have now.
My main question is how are we going to allocated when we have an over actuated system with motors and control surfaces.
My suggestion would be during hover and transition to first allocated the motors and then use the control surfaces to take care of the un-allocated input, should there be any.
The effect would be that the control surfaces will only start to move once the motors start to saturate.
My concern with using both motors and actuators at the same time, even if the motors are not saturated, is that especially during hover and transition airspeed might not be well known and thus, it's not entirely clear how much torque the control surfaces really generated.

Reply:
> I don't see a reason why we could not.

 Cool, let's do it then!

> The effect would be that the control surfaces will only start to move once the motors start to saturate.

IMO waiting for a subset of actuators to saturate before using the full capability of the actuators is not clean, maybe even dangerous. 

> My main question is how are we going to allocate when we have an over actuated system with motors and control surfaces.

That is what pseudoinverse is for: it computes the smallest combination of actuator movement that generates the desired thrust&torque. Say you scale flap effectiveness with airspeed, then during hover flaps will be so ineffective that almost all control will be allocated to the rotors.

Reply:
@jlecoeur I agree, but still if your airspeed is not very accurate then you might still run into issues.
I think we will just have to test it in real.

Reply:
You are right, it depends on the quality of estimated airspeed. What does it look like today with a VTOL hovering in low wing? Do we have an estimate down to 0m/s?

Reply:
@jlecoeur Depends a bit which sensor. The sensirion ones (sdp33x) are quite good and I think measured down to 0. The older ones (ms5611) become accurate at about 8m/s or so.